crates/gitbutler-stack/src/stack_branch.rs

```diff
@@ -74,31 +79,136 @@ impl RepositoryExt for git2::Repository {
 }
 
 impl StackBranch {
-    pub fn new(head: CommitOrChangeId, name: String, description: Option<String>) -> Self {
-        StackBranch {
+    pub fn new(
+        head: CommitOrChangeId,
+        name: String,
+        description: Option<String>,
+        repo: &gix::Repository,
+    ) -> Result<Self> {
+        let branch = StackBranch {
             head,
             name,
             description,
             pr_number: None,
             archived: false,
             review_id: None,
-        }
+        };
+        branch.set_real_reference(repo, &branch.head)?;
+        Ok(branch)
     }
 
     pub fn head(&self) -> &CommitOrChangeId {
         &self.head
     }
 
-    pub fn set_head(&mut self, head: CommitOrChangeId) {
+    /// This will update the commit that this points to (the virtual reference in virtual_branches.toml) as well as update of create a real git reference.
+    /// If this points to a change id, it's a noop operation. In practice, moving forward, new CommitOrChangeId entries will always be CommitId and ChangeId may only appear in deserialized data.
+    pub fn set_head(
+        &mut self,
+        head: CommitOrChangeId,
+        repo: &gix::Repository,
+    ) -> Result<Option<BString>> {
+        let refname = self.set_real_reference(repo, &head)?;
         self.head = head;
+        Ok(refname)
     }
 
     pub fn name(&self) -> &String {
         &self.name
     }
 
-    pub fn set_name(&mut self, name: String) {
+    pub fn set_name(&mut self, name: String, repo: &gix::Repository) -> Result<()> {
+        self.rename_real_reference(&name, repo)?;
         self.name = name;
+        Ok(())
+    }
+
+    pub fn delete_reference(&self, repo: &gix::Repository) -> Result<()> {
+        let oid = match self.head.clone() {
+            CommitOrChangeId::CommitId(id) => gix::ObjectId::from_str(&id)?,
+            CommitOrChangeId::ChangeId(_) => return Ok(()), // noop
+        };
+        let current_name: BString = qualified_reference_name(self.name()).into();
+        if let Some(reference) = repo.try_find_reference(&current_name)? {
+            let delete = RefEdit {
+                change: Change::Delete {
+                    expected: PreviousValue::MustExistAndMatch(oid.into()),
+                    log: RefLog::AndReference,
+                },
+                name: reference.name().into(),
+                deref: false,
+            };
+            repo.edit_references([delete])?;
+        }
+        Ok(())
+    }
+
+    fn rename_real_reference(&self, name: &str, repo: &gix::Repository) -> Result<()> {
+        let current_name: BString = qualified_reference_name(self.name()).into();
```

Comment:
If the current name was equal to the new name, the operation would probably fail as the transaction would try to obtain a lock for each ref-edit.
Thus the caller would have to assure the name is different or face an error.

Reply:
Oh i see! I am adding a noop return if the name is the same, just to avoid this codepath alltogether